R/bs-theme.R

```diff
@@ -1,3 +1,158 @@
+#'   actions, and to show active selection state in some Bootstrap components.
+#'   Generally a bold, saturated color that contrasts with the theme's base
+#'   colors.
+#' @param secondary A color for components and messages that don't need to stand
+#'   out. (Not supported in Bootstrap 3.)
+#' @param success A color for messages that indicate an operation has succeeded.
+#'   Typically green.
+#' @param info A color for messages that are informative but not critical. Typically a
+#'   shade of blue-green.
+#' @param warning A color for warning messages. Typically yellow.
+#' @param danger A color for errors. Typically red.
+#' @param base_font The default typeface.
+#' @param code_font The typeface to be used for code. Be sure this is monospace!
+#' @param heading_font The typeface to be used for heading elements.
+#'
+#' @references \url{https://getbootstrap.com/docs/4.4/getting-started/theming/}
+#' @references \url{https://rstudio.github.io/sass/}
+#'
+#' @export
+#' @examples
+#'
+#' theme <- bs_theme(
+#'   # Controls the default grayscale palette
+#'   bg = "#202123", fg = "#B8BCC2",
+#'   # Controls the accent (e.g., hyperlink, button, etc) colors
+#'   primary = "#EA80FC", secondary = "#48DAC6",
+#'   base_font = c("Grandstander", "sans-serif"),
+#'   code_font = c("Courier", "monospace"),
+#'   heading_font = "'Helvetica Neue', Helvetica, sans-serif",
+#'   # Can also add lower-level customization
+#'   "input-border-color" = "#EA80FC"
+#' )
+#' if (interactive()) {
+#'   bs_theme_preview(theme)
+#' }
+#'
+#' @export
+bs_theme <- function(version = version_default(), bootswatch = NULL, ...,
+                     bg = NULL, fg = NULL, primary = NULL, secondary = NULL,
+                     success = NULL, info = NULL, warning = NULL, danger = NULL,
+                     base_font = NULL, code_font = NULL, heading_font = NULL) {
+  version <- version_resolve(version)
+  bootswatch <- bootswatch_theme_resolve(bootswatch, version)
+  theme <- bs_add_layers(
+    add_class(sass_layer(), "bs_theme"),
+    bootstrap_layer(version),
+    if (identical(version, "4+3")) bs3compat_layer(),
+    # This will set a $navbar-height Sass var, even if no bootswatch is used
+    # TODO: maybe make navbar adjustment via jQuery instead? https://stackoverflow.com/a/46021836/1583084
+    navbar_height_layer(bootswatch, version),
+    bootswatch_layer(bootswatch, version)
+  )
+  bs_theme_update(theme)
+}
+
+#' @rdname bs_theme
+#' @export
+bs_theme_update <- function(theme = bs_theme(), ..., bg = NULL, fg = NULL,
+                            primary = NULL,  secondary = NULL, success = NULL,
+                            info = NULL, warning = NULL, danger = NULL,
+                            base_font = NULL, code_font = NULL, heading_font = NULL) {
+  theme <- assert_bs_theme(theme)
+  # See R/bs-theme-update.R for the implementation of these
+  theme <- bs_base_colors(theme, bg = bg, fg = fg)
```

Comment:
What should happen in the case of `bs_theme_update(bs_theme(bg = "black", fg = "white"), fg = "red")` (should we try to remember the previous `bg`)?

Make sure we also have ample tests of this logic